test/extensions/filters/network/proxy_protocol/proxy_protocol_integration_test.cc

```diff
@@ -0,0 +1,130 @@
+#include "envoy/extensions/filters/network/proxy_protocol/v3/proxy_protocol.pb.h"
+
+#include "extensions/filters/common/proxy_protocol/proxy_protocol_header.h"
+
+#include "test/integration/integration.h"
+
+using envoy::extensions::filters::network::proxy_protocol::v3::ProxyProtocol_Version;
+
+namespace Envoy {
+namespace {
+
+class ProxyProtocolIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                     public BaseIntegrationTest {
+public:
+  ProxyProtocolIntegrationTest()
+      : BaseIntegrationTest(GetParam(), ConfigHelper::TCP_PROXY_CONFIG) {}
+
+  ~ProxyProtocolIntegrationTest() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void initialize() override {
+    enable_half_close_ = true;
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      auto* cluster_0 = bootstrap.mutable_static_resources()->mutable_clusters(0);
+      auto* filter = cluster_0->add_filters();
+      filter->set_name("envoy.filters.network.proxy_protocol");
```

Comment:
What would you think of a test where the client sent its own proxy protocol header, and we make sure that if we have the downstream proxy protocol header, that we emit a proxy protocol header based on the one we read in?

Basically if we have a multi-layer set up that Envoy can remove and add correctly?

Reply:
The intent of the feature was to have Envoy generate the header for clients that don't send the header themselves. Currently it doesn't do any processing to check if the header is being sent already. We could modify the filter to check if the header was sent, and then only generate the header if the client didn't pass it. This would however mean the downstreams would need to be trusted, because they could send whatever address they want in the header.